### DISCUSSION TITLE
Lesson 31: failing testFulfillRandomWordsPicksAWinnerResetsAndSendsMoney() test

### DISCUSSION THREAD

**Discussion by Swarnim-Chandve** [Cyfrin/foundry-full-course-cu#1468] [Q&A]
here's the code: 
RaffleTest.sol : 
```// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity ^0.8.18;

import {DeployRaffle} from "../../script/DeployRaffle.s.sol";
import {Raffle} from "../../src/Raffle.sol";
import {Test,console} from "forge-std/Test.sol";
import {HelperConfig} from "../../script/HelperConfig.s.sol";
import {Vm} from "forge-std/Vm.sol";
import {VRFCoordinatorV2Mock} from "@chainlink/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol";

contract RaffleTest is Test {

    //Events
    event EnteredRaffle(address indexed player);


    Raffle raffle;
    HelperConfig helperConfig;
         uint256 entranceFee;
        uint256 interval;
        address vrfCoordinator;
        bytes32 gasLane;
        uint64 subscriptionId;
        uint32 callbackGasLimit;
        address link;
        

    address public PLAYER = makeAddr("player");
    uint256 public constant STARTING_USER_BALANCE = 10 ether;

    function setUp() external {
        DeployRaffle deployer = new DeployRaffle();
        (raffle, helperConfig) = deployer.run();
        (
        entranceFee,
         interval,
         vrfCoordinator,
         gasLane,
         subscriptionId,
         callbackGasLimit,
         link

        )= helperConfig.activeNetworkConfig();
    
    vm.deal(PLAYER, STARTING_USER_BALANCE);
    }

    function testRaffleInitializesInOpenState() public view {
        assert(raffle.getRaffleState() == Raffle.RaffleState.OPEN );
    }

    //enterRaffle

    function testRaffleRevertsWhenYouDontPayEnough() public {
        vm.prank(PLAYER);

        vm.expectRevert(Raffle.Raffle__NotEnoughEthSent.selector);
        raffle.enterRaffle();
    }


    function testRaffleRecordsPlayerWhenTheyEnter() public {
        vm.prank(PLAYER);
        raffle.enterRaffle{value: entranceFee}();
        address playerRecorded = raffle.getPlayer(0);
        assert(playerRecorded == PLAYER);
    
    }

    function testEmitsEventOnEntrance() public {
        vm.prank(PLAYER);
        vm.expectEmit(true, false, false, false, address(raffle));
        emit EnteredRaffle(PLAYER);
        raffle.enterRaffle{value: entranceFee}();
    }

    function testCanEnterWhenRaffleIsCalculating() public {
        vm.prank(PLAYER);
        raffle.enterRaffle{value: entranceFee}();
        vm.warp(block.timestamp + interval + 1);
        vm.roll(block.number + 1);
        raffle.performUpkeep("");

        vm.expectRevert(Raffle.Raffle_RaffleNotOpen.selector);
        vm.prank(PLAYER);
        raffle.enterRaffle{value: entranceFee}();
    }



    //checkUpkeep

    function testCheckUpkeepReturnsFalseIfIthasNoBalance() public {
    vm.warp(block.timestamp + interval + 1);
    vm.roll(block.number + 1);


    //Act
    (bool upkeepNeeded, ) = raffle.checkUpkeep("");

    //Assert
    assert(!upkeepNeeded);
    }

    function testCheckUpkeepReturnsFalseIfRaffleNotOpen() public {
        //Arrange
        vm.prank(PLAYER);
        raffle.enterRaffle{value: entranceFee}();
        vm.warp(block.timestamp + interval + 1);
        vm.roll(block.number + 1);
        raffle.performUpkeep("");


        //Act
        (bool upkeepNeeded, ) = raffle.checkUpkeep("");
        //Assert
        assert(upkeepNeeded == false);

    }


        //testCheckUpkeepReturnsFalseIfEnoughTimeHasntPassed
        //testCheckUpkeepReturnsTrueWhenParametersAreGood


        function testPerformUpkeepCanOnlyRunIfCheckUpkeepIsTrue() public {
            //Arrange
            vm.prank(PLAYER);
            raffle.enterRaffle{value: entranceFee}();
            vm.warp(block.timestamp + interval + 1);
            vm.roll(block.number + 1);


            //Act / Assert
            raffle.performUpkeep("");
        }


        function testPerformUpkeepRevertsIfCheckUpkeepIfFalse() public {
            //Arrange
            uint256 currentBalance = 0;
            uint256 numPlayers = 0;
            uint256 raffleState = 0;


            // Act/Assert
            vm.expectRevert(
                
                abi.encodeWithSelector(
                    Raffle.Raffle_UpkeepNotNeeded.selector,
                    currentBalance,
                    numPlayers,
                    raffleState
                )
            );
            raffle.performUpkeep("");
        }

        modifier raffleEnteredAndTimePassed(){
            vm.prank(PLAYER);
            raffle.enterRaffle{value: entranceFee}();
            vm.warp(block.timestamp + interval + 1);
            vm.roll(block.number + 1);
            _;
        }

        function testPerformUpkeepUpdatesRaffleStateAndEmitsRequestId()
         public raffleEnteredAndTimePassed {
          
          //Act
          vm.recordLogs();
          raffle.performUpkeep("");
          Vm.Log[] memory entries = vm.getRecordedLogs();
          bytes32 requestId = entries[1].topics[1];

        Raffle.RaffleState rstate = raffle.getRaffleState();

          assert(uint256(requestId) > 0);
          assert(uint256(rstate) == 1); 
        }

    function testFulfillRandomWordsCanOnlyBeCalledAfterPerformUpkeep(uint256 randomRequestId) public raffleEnteredAndTimePassed {
        //Arrange
        vm.expectRevert("nonexistent request");
        VRFCoordinatorV2Mock(vrfCoordinator).fulfillRandomWords(randomRequestId, address(raffle));

    }

    // function testFulfillRandomWordsPicksAWinnerResetsAndSendsMoney() public raffleEnteredAndTimePassed{
    //     uint256 additionalEntrants = 5;
    //     uint256 startingIndex = 1;
    //     for(uint256 i = startingIndex ;i< startingIndex+ additionalEntrants;i++){
    //         address player = address(uint160(i));
    //         hoax(player, STARTING_USER_BALANCE);
    //         raffle.enterRaffle{value : entranceFee};
    //     }

    //     uint256 prize = entranceFee * (additionalEntrants + 1);

    //     vm.recordLogs();
    //     raffle.performUpkeep("");
    //     Vm.Log[] memory entries = vm.getRecordedLogs();
    //     bytes32 requestId = entries[1].topics[1];

    //     uint256 previousTimeStamp = raffle.getLastTimeStamp();

    //     VRFCoordinatorV2Mock(vrfCoordinator).fulfillRandomWords(
    //         uint256(requestId),
    //         address(raffle)
    //     );

    //     // assert(uint256(raffle.getRaffleState()) == 0);
    //     // assert(raffle.getRecentWinner != address(0));
    //     // assert(raffle.getLengthOfPlayers() == 0);
    //     // assert(previousTimeStamp < raffle.getLastTimeStamp());
    //     assert(raffle.getRecentWinner().balance == STARTING_USER_BALANCE + prize);
    // }

    function testFulfillRandomWordsPicksAWinnerResetsAndSendsMoney() public raffleEnteredAndTimePassed {
    // Mock the behavior or state of the players using the prank function
    vm.prank(PLAYER);
    
    uint256 additionalEntrants = 5;
    uint256 startingIndex = 1;
    
    for (uint256 i = startingIndex; i < startingIndex + additionalEntrants; i++) {
        address player = address(uint160(i));
        vm.prank(player);
        raffle.enterRaffle{value: entranceFee}();
    }

    // Record the state before calling performUpkeep
    uint256 previousTimeStamp = raffle.getLastTimeStamp();
    uint256 initialBalance = raffle.getRecentWinner().balance;

    // Perform the upkeep and fulfill the random words
    vm.recordLogs();
    raffle.performUpkeep("");
    Vm.Log[] memory entries = vm.getRecordedLogs();
    bytes32 requestId = entries[1].topics[1];

    VRFCoordinatorV2Mock(vrfCoordinator).fulfillRandomWords(
        uint256(requestId),
        address(raffle)
    );

    // Assertions after the randomness is fulfilled
    // assert(raffle.getRaffleState() == Raffle.RaffleState.OPEN);
    // assert(raffle.getRecentWinner() != address(0));
    // assert(raffle.getLengthOfPlayers() == 0);
    // assert(previousTimeStamp < raffle.getLastTimeStamp());

    // Calculate the expected prize
    uint256 prize = entranceFee * (additionalEntrants + 1);

    // Check if the recent winner received the expected prize
    assert(raffle.getRecentWinner().balance == STARTING_USER_BALANCE + prize - entranceFee);
}

}```

**Comment by Swarnim-Chandve** [comment]
> here's the code: RaffleTest.sol :
> 
> ```
> pragma solidity ^0.8.18;
> 
> import {DeployRaffle} from "../../script/DeployRaffle.s.sol";
> import {Raffle} from "../../src/Raffle.sol";
> import {Test,console} from "forge-std/Test.sol";
> import {HelperConfig} from "../../script/HelperConfig.s.sol";
> import {Vm} from "forge-std/Vm.sol";
> import {VRFCoordinatorV2Mock} from "@chainlink/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol";
> 
> contract RaffleTest is Test {
> 
>     //Events
>     event EnteredRaffle(address indexed player);
> 
> 
>     Raffle raffle;
>     HelperConfig helperConfig;
>          uint256 entranceFee;
>         uint256 interval;
>         address vrfCoordinator;
>         bytes32 gasLane;
>         uint64 subscriptionId;
>         uint32 callbackGasLimit;
>         address link;
>         
> 
>     address public PLAYER = makeAddr("player");
>     uint256 public constant STARTING_USER_BALANCE = 10 ether;
> 
>     function setUp() external {
>         DeployRaffle deployer = new DeployRaffle();
>         (raffle, helperConfig) = deployer.run();
>         (
>         entranceFee,
>          interval,
>          vrfCoordinator,
>          gasLane,
>          subscriptionId,
>          callbackGasLimit,
>          link
> 
>         )= helperConfig.activeNetworkConfig();
>     
>     vm.deal(PLAYER, STARTING_USER_BALANCE);
>     }
> 
>     function testRaffleInitializesInOpenState() public view {
>         assert(raffle.getRaffleState() == Raffle.RaffleState.OPEN );
>     }
> 
>     //enterRaffle
> 
>     function testRaffleRevertsWhenYouDontPayEnough() public {
>         vm.prank(PLAYER);
> 
>         vm.expectRevert(Raffle.Raffle__NotEnoughEthSent.selector);
>         raffle.enterRaffle();
>     }
> 
> 
>     function testRaffleRecordsPlayerWhenTheyEnter() public {
>         vm.prank(PLAYER);
>         raffle.enterRaffle{value: entranceFee}();
>         address playerRecorded = raffle.getPlayer(0);
>         assert(playerRecorded == PLAYER);
>     
>     }
> 
>     function testEmitsEventOnEntrance() public {
>         vm.prank(PLAYER);
>         vm.expectEmit(true, false, false, false, address(raffle));
>         emit EnteredRaffle(PLAYER);
>         raffle.enterRaffle{value: entranceFee}();
>     }
> 
>     function testCanEnterWhenRaffleIsCalculating() public {
>         vm.prank(PLAYER);
>         raffle.enterRaffle{value: entranceFee}();
>         vm.warp(block.timestamp + interval + 1);
>         vm.roll(block.number + 1);
>         raffle.performUpkeep("");
> 
>         vm.expectRevert(Raffle.Raffle_RaffleNotOpen.selector);
>         vm.prank(PLAYER);
>         raffle.enterRaffle{value: entranceFee}();
>     }
> 
> 
> 
>     //checkUpkeep
> 
>     function testCheckUpkeepReturnsFalseIfIthasNoBalance() public {
>     vm.warp(block.timestamp + interval + 1);
>     vm.roll(block.number + 1);
> 
> 
>     //Act
>     (bool upkeepNeeded, ) = raffle.checkUpkeep("");
> 
>     //Assert
>     assert(!upkeepNeeded);
>     }
> 
>     function testCheckUpkeepReturnsFalseIfRaffleNotOpen() public {
>         //Arrange
>         vm.prank(PLAYER);
>         raffle.enterRaffle{value: entranceFee}();
>         vm.warp(block.timestamp + interval + 1);
>         vm.roll(block.number + 1);
>         raffle.performUpkeep("");
> 
> 
>         //Act
>         (bool upkeepNeeded, ) = raffle.checkUpkeep("");
>         //Assert
>         assert(upkeepNeeded == false);
> 
>     }
> 
> 
>         //testCheckUpkeepReturnsFalseIfEnoughTimeHasntPassed
>         //testCheckUpkeepReturnsTrueWhenParametersAreGood
> 
> 
>         function testPerformUpkeepCanOnlyRunIfCheckUpkeepIsTrue() public {
>             //Arrange
>             vm.prank(PLAYER);
>             raffle.enterRaffle{value: entranceFee}();
>             vm.warp(block.timestamp + interval + 1);
>             vm.roll(block.number + 1);
> 
> 
>             //Act / Assert
>             raffle.performUpkeep("");
>         }
> 
> 
>         function testPerformUpkeepRevertsIfCheckUpkeepIfFalse() public {
>             //Arrange
>             uint256 currentBalance = 0;
>             uint256 numPlayers = 0;
>             uint256 raffleState = 0;
> 
> 
>             // Act/Assert
>             vm.expectRevert(
>                 
>                 abi.encodeWithSelector(
>                     Raffle.Raffle_UpkeepNotNeeded.selector,
>                     currentBalance,
>                     numPlayers,
>                     raffleState
>                 )
>             );
>             raffle.performUpkeep("");
>         }
> 
>         modifier raffleEnteredAndTimePassed(){
>             vm.prank(PLAYER);
>             raffle.enterRaffle{value: entranceFee}();
>             vm.warp(block.timestamp + interval + 1);
>             vm.roll(block.number + 1);
>             _;
>         }
> 
>         function testPerformUpkeepUpdatesRaffleStateAndEmitsRequestId()
>          public raffleEnteredAndTimePassed {
>           
>           //Act
>           vm.recordLogs();
>           raffle.performUpkeep("");
>           Vm.Log[] memory entries = vm.getRecordedLogs();
>           bytes32 requestId = entries[1].topics[1];
> 
>         Raffle.RaffleState rstate = raffle.getRaffleState();
> 
>           assert(uint256(requestId) > 0);
>           assert(uint256(rstate) == 1); 
>         }
> 
>     function testFulfillRandomWordsCanOnlyBeCalledAfterPerformUpkeep(uint256 randomRequestId) public raffleEnteredAndTimePassed {
>         //Arrange
>         vm.expectRevert("nonexistent request");
>         VRFCoordinatorV2Mock(vrfCoordinator).fulfillRandomWords(randomRequestId, address(raffle));
> 
>     }
> 
>     // function testFulfillRandomWordsPicksAWinnerResetsAndSendsMoney() public raffleEnteredAndTimePassed{
>     //     uint256 additionalEntrants = 5;
>     //     uint256 startingIndex = 1;
>     //     for(uint256 i = startingIndex ;i< startingIndex+ additionalEntrants;i++){
>     //         address player = address(uint160(i));
>     //         hoax(player, STARTING_USER_BALANCE);
>     //         raffle.enterRaffle{value : entranceFee};
>     //     }
> 
>     //     uint256 prize = entranceFee * (additionalEntrants + 1);
> 
>     //     vm.recordLogs();
>     //     raffle.performUpkeep("");
>     //     Vm.Log[] memory entries = vm.getRecordedLogs();
>     //     bytes32 requestId = entries[1].topics[1];
> 
>     //     uint256 previousTimeStamp = raffle.getLastTimeStamp();
> 
>     //     VRFCoordinatorV2Mock(vrfCoordinator).fulfillRandomWords(
>     //         uint256(requestId),
>     //         address(raffle)
>     //     );
> 
>     //     // assert(uint256(raffle.getRaffleState()) == 0);
>     //     // assert(raffle.getRecentWinner != address(0));
>     //     // assert(raffle.getLengthOfPlayers() == 0);
>     //     // assert(previousTimeStamp < raffle.getLastTimeStamp());
>     //     assert(raffle.getRecentWinner().balance == STARTING_USER_BALANCE + prize);
>     // }
> 
>     function testFulfillRandomWordsPicksAWinnerResetsAndSendsMoney() public raffleEnteredAndTimePassed {
>     // Mock the behavior or state of the players using the prank function
>     vm.prank(PLAYER);
>     
>     uint256 additionalEntrants = 5;
>     uint256 startingIndex = 1;
>     
>     for (uint256 i = startingIndex; i < startingIndex + additionalEntrants; i++) {
>         address player = address(uint160(i));
>         vm.prank(player);
>         raffle.enterRaffle{value: entranceFee}();
>     }
> 
>     // Record the state before calling performUpkeep
>     uint256 previousTimeStamp = raffle.getLastTimeStamp();
>     uint256 initialBalance = raffle.getRecentWinner().balance;
> 
>     // Perform the upkeep and fulfill the random words
>     vm.recordLogs();
>     raffle.performUpkeep("");
>     Vm.Log[] memory entries = vm.getRecordedLogs();
>     bytes32 requestId = entries[1].topics[1];
> 
>     VRFCoordinatorV2Mock(vrfCoordinator).fulfillRandomWords(
>         uint256(requestId),
>         address(raffle)
>     );
> 
>     // Assertions after the randomness is fulfilled
>     // assert(raffle.getRaffleState() == Raffle.RaffleState.OPEN);
>     // assert(raffle.getRecentWinner() != address(0));
>     // assert(raffle.getLengthOfPlayers() == 0);
>     // assert(previousTimeStamp < raffle.getLastTimeStamp());
> 
>     // Calculate the expected prize
>     uint256 prize = entranceFee * (additionalEntrants + 1);
> 
>     // Check if the recent winner received the expected prize
>     assert(raffle.getRecentWinner().balance == STARTING_USER_BALANCE + prize - entranceFee);
> }
> 
> }```


here's the error: ```
  [71010] RaffleTest::testFulfillRandomWordsPicksAWinnerResetsAndSendsMoney()
    ├─ [0] VM::prank(player: [0x44E97aF4418b7a17AABD8090bEA0A471a366305C])
    │   └─ ← ()
    ├─ [47823] Raffle::enterRaffle{value: 10000000000000000}()
    │   ├─ emit EnteredRaffle(player: player: [0x44E97aF4418b7a17AABD8090bEA0A471a366305C])
    │   └─ ← ()
    ├─ [0] VM::warp(32)
    │   └─ ← ()
    ├─ [0] VM::roll(2)
    │   └─ ← ()
    ├─ [0] VM::prank(player: [0x44E97aF4418b7a17AABD8090bEA0A471a366305C])
    │   └─ ← ()
    ├─ [0] VM::prank(0x0000000000000000000000000000000000000001)
    │   └─ ← cannot overwrite a prank until it is applied at least once
    └─ ← cannot overwrite a prank until it is applied at least once

Test result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 29.67ms
 
Ran 1 test suites: 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/unit/RaffleTest.t.sol:RaffleTest
[FAIL. Reason: cannot overwrite a prank until it is applied at least once] testFulfillRandomWordsPicksAWinnerResetsAndSendsMoney() (gas: 71010)

Encountered a total of 1 failing tests, 0 tests succeeded
```

**Reply by EngrPips** [reply]
Hello @Swarnim-Chandve , So what happened is that in your `testFulfillRandomWordsPicksAWinnerResetsAndSendsMoney` test the first line inside that function pranked an address `PLAYER` so that the next transaction that would happened the `PLAYER` address would be the `msg.sender` but you didn't actually make any transaction till inside the for loop where you have another prank that pranked an address `player` which doesn't exist in your test file. Lets remember that you had initially pranked `PLAYER` and you didn't make any transaction and now you try to make another prank to prank `player` so foundry complain that you didn't actually use your first prank for any transaction and you are trying to start another so it revert. Just to be clear why we you have `VM::prank(0x0000000000000000000000000000000000000001)` in the stack trace of your test ouput is that the address `player` you tried to prank doesn't actually exist in your test file so foundry defaulted it to `0x0000000000000000000000000000000000000001`

To solve your problem , remove the first `vm.prank()` in your `testFulfillRandomWordsPicksAWinnerResetsAndSendsMoney` function and correct the `vm.prank()` in the for loop to prank the correct address which is `PLAYER`.

I hope i explained enough to make things clear to you, please let me know if anything is confusing
